File: src/main/java/com/neo/caption/ocr/service/impl/TesseractServiceImpl.java
package com.neo.caption.ocr.service.impl;

import com.google.common.base.Splitter;
import com.neo.caption.ocr.domain.dto.TessLanguageDto;
import com.neo.caption.ocr.property.TesseractProperties;
import com.neo.caption.ocr.service.TesseractService;
import com.neo.caption.ocr.util.BaseUtil;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.bytedeco.tesseract.StringVector;
import org.bytedeco.tesseract.TessBaseAPI;
import org.springframework.stereotype.Service;

import javax.annotation.PostConstruct;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

@Service
@Slf4j
@RequiredArgsConstructor
public class TesseractServiceImpl implements TesseractService {

    private final TesseractProperties tesseractProperties;
    private final Splitter commaSplitter;

    private Path tessDataPath;
    private Path tessConfigPath;
    private List<TessLanguageDto> tessLanguageDtoList;

    @PostConstruct
    public void init() {
        this.tessDataPath = Paths.get("").resolve("app").resolve("tessdata");
        if (!Files.exists(tessDataPath)) {
            log.warn("The tessdata does not exist: {}", tessDataPath.toAbsolutePath());
            this.tessLanguageDtoList = Collections.emptyList();
            return;
        }
        this.tessConfigPath = tessDataPath.resolve("config");
        var languageList = commaSplitter.splitToList(tesseractProperties.getSupportedLanguage());
        if (languageList.isEmpty()) {
            log.warn("not set the supported languages");
            this.tessLanguageDtoList = Collections.emptyList();
            return;
        }
        this.tessLanguageDtoList = new ArrayList<>(languageList.size());
        languageList.forEach(this::addToTessLanguageDtoList);
    }

    @Override
    public TessBaseAPI initTessBaseApi(String language) {
        try (var tessBaseAPI = new TessBaseAPI();
             var variableName = new StringVector();
             var variableValue = new StringVector()) {
            for (var config : tesseractProperties.getConfigs()) {
                variableName.put(config.getName());
                variableValue.put(config.getValue());
            }
            variableName.put("tessedit_pageseg_mode");
            variableValue.put(BaseUtil.v2s(tesseractProperties.getPageSegMode()));
            tessBaseAPI.Init(tessDataPath.toAbsolutePath().toString(),
                    language,
                    tesseractProperties.getOcrEngineMode(),
                    new byte[0],
                    (int) variableName.size(),
                    variableName,
                    variableValue,
                    true);
            if (Files.exists(tessConfigPath)) {
                tessBaseAPI.ReadConfigFile(tessConfigPath.toAbsolutePath().toString());
            }
            return tessBaseAPI;
        }
    }

    @Override
    public List<TessLanguageDto> getTessLanguageDtoList() {
        return tessLanguageDtoList;
    }

    private void addToTessLanguageDtoList(String language) {
        var dataPath = tessDataPath.resolve(language + ".traineddata");
        var dto = new TessLanguageDto(language, Files.exists(dataPath));
        tessLanguageDtoList.add(dto);
    }

}

File: src/main/kotlin/com/neo/caption/ocr/service/impl/TesseractServiceImpl.kt
package com.neo.caption.ocr.service.impl

import com.neo.caption.ocr.annotation.Slf4j
import com.neo.caption.ocr.annotation.Slf4j.Companion.log
import com.neo.caption.ocr.constant.OCREngineModeEnum
import com.neo.caption.ocr.constant.PageSegModeEnum
import com.neo.caption.ocr.domain.entity.TesseractConfig
import com.neo.caption.ocr.domain.vo.TesseractConfigVo
import com.neo.caption.ocr.domain.vo.TesseractOptionVo
import com.neo.caption.ocr.property.TesseractProperties
import com.neo.caption.ocr.service.TesseractService
import org.bytedeco.tesseract.TessBaseAPI
import org.springframework.cache.annotation.CacheConfig
import org.springframework.cache.annotation.Cacheable
import org.springframework.stereotype.Service
import java.nio.file.Files
import java.nio.file.Path
import kotlin.io.path.absolutePathString
import kotlin.io.path.name
import kotlin.io.path.nameWithoutExtension
import kotlin.io.path.notExists

@Slf4j
@Service
@CacheConfig(cacheNames = ["tess::"])
class TesseractServiceImpl(
    private val tesseractProperties: TesseractProperties,
) : TesseractService {

    private val tessdataPath = Path.of("").resolve("lib").resolve("tessdata")
    override val supportedLanguage: List<String> by lazy {
        if (tessdataPath.notExists()) {
            log.warn { "not found tessdata in ${tessdataPath.absolutePathString()}" }
            emptyList()
        } else {
            Files.walk(tessdataPath)
                .filter { it.name.endsWith(".traineddata") }
                .map { it.nameWithoutExtension }
                .toList()
        }
    }

    @Cacheable("default-config")
    override fun getDefaultConfig(): TesseractConfigVo {
        return TesseractConfigVo(
            tesseractProperties.ocrEngineMode,
            tesseractProperties.pageSegMode,
            tesseractProperties.language,
            tesseractProperties.linkedHashMap()
        )
    }

    @Cacheable(key = "'option'")
    override fun getTesseractOption() =
        TesseractOptionVo(
            oemModeList = OCREngineModeEnum.values().toList(),
            psmModeList = PageSegModeEnum.values().toList(),
            supportedLanguage = supportedLanguage,
            vectorKeyList = emptyList()
        )

    override fun initTessBaseApi(tesseractConfig: TesseractConfig): TessBaseAPI {
        return TessBaseAPI().apply {
            this.Init(
                tessdataPath.absolutePathString(),
                tesseractConfig.language,
                tesseractConfig.oem,
                ByteArray(0),
                tesseractConfig.vectorKey.size().toInt(),
                tesseractConfig.vectorKey,
                tesseractConfig.vectorValue,
                true
            )
        }
    }

}
